client/app/components/items-list/hooks/useItemsListExtraActions.js

```diff
@@ -0,0 +1,74 @@
+import { filter, includes, intersection } from "lodash";
+import React, { useState, useMemo, useEffect, useCallback } from "react";
+import Checkbox from "antd/lib/checkbox";
+import { Columns } from "../components/ItemsTable";
+
+export default function useItemsListExtraActions(controller, listColumns, ExtraActionsComponent) {
+  const [actionsState, setActionsState] = useState({ isAvailable: false });
+  const [selectedItems, setSelectedItems] = useState([]);
+
+  // clear selection when page changes
+  useEffect(() => {
+    setSelectedItems([]);
+  }, [controller.pageItems, actionsState.isAvailable]);
+
+  const areAllItemsSelected = useMemo(() => {
+    const allItems = controller.pageItems;
+    if (allItems.length === 0 || selectedItems.length === 0) {
+      return false;
+    }
+    return intersection(selectedItems, allItems).length === allItems.length;
+  }, [selectedItems, controller.pageItems]);
+
+  const toggleAllItems = useCallback(() => {
+    if (areAllItemsSelected) {
+      setSelectedItems([]);
+    } else {
+      setSelectedItems(controller.pageItems);
+    }
+  }, [areAllItemsSelected, controller.pageItems]);
+
+  const toggleItem = useCallback(
+    item => {
+      if (includes(selectedItems, item)) {
+        setSelectedItems(filter(selectedItems, s => s !== item));
+      } else {
+        setSelectedItems([...selectedItems, item]);
+      }
+    },
+    [selectedItems]
+  );
+
+  const checkboxColumn = useMemo(() => {
+    return Columns.custom(
+      (text, item) => <Checkbox checked={includes(selectedItems, item)} onChange={() => toggleItem(item)} />,
+      {
+        title: () => <Checkbox checked={areAllItemsSelected} onChange={toggleAllItems} />,
+        field: "id",
+        width: "1%",
+      }
+    );
+  }, [selectedItems, areAllItemsSelected, toggleAllItems, toggleItem]);
+
+  const Component = useCallback(
+    function ItemsListExtraActionsComponentWrapper(props) {
+      // this check mostly needed to avoid eslint exhaustive deps warning
+      if (!ExtraActionsComponent) {
+        return null;
+      }
+
+      return <ExtraActionsComponent onStateChange={setActionsState} {...props} />;
+    },
+    [ExtraActionsComponent]
+  );
+
+  return useMemo(() => {
+    return {
+      areExtraActionsAvailable: actionsState.isAvailable,
+      listColumns: actionsState.isAvailable ? [checkboxColumn, ...listColumns] : listColumns,
+      Component,
+      selectedItems,
+      setSelectedItems,
+    };
+  }, [actionsState, listColumns, checkboxColumn, selectedItems, Component]);
```

Comment:
This `useMemo` is kinda not necessary as all props are read destructured anyway. Except for the `listColumns` as it creates a new array, so it would require a new one only for it. Doing this later makes it a bit better to guarantee it will only update `listColumns` when it needs, but it's not that relevant here anyway.

---

BTW, you are not a big fan of returning things directly in the arrow func? 😅 Both here and in the `useMemo` for `checkboxColumn`

Reply:
I prefer to keep useMemo in such cases just to avoid surprises when using hooks.

> you are not a big fan of returning things directly in the arrow func?

There was other code, which I later removed :man_facepalming: Will fix